Python/pylifecycle.c

```diff
@@ -2719,6 +2729,70 @@ init_sys_streams(PyThreadState *tstate)
 }
 
 
+#ifdef __ANDROID__
+#include <android/log.h>
+
+static PyObject *
+android_log_write_impl(PyObject *self, PyObject *args)
+{
+    int prio = 0;
+    char *tag = NULL;
+    char *text = NULL;
+    Py_ssize_t *text_len = 0;
+    if (!PyArg_ParseTuple(args, "iss#", &prio, &tag, &text, &text_len))
+        return NULL;
+
+    // Despite its name, this function is part of the public API
+    // (https://developer.android.com/ndk/reference/group/logging).
+    __android_log_write(prio, tag, text);
+    Py_RETURN_NONE;
+}
+
+
+static PyMethodDef android_log_write_method = {
+    "android_log_write", android_log_write_impl, METH_VARARGS
+};
+
+
+static PyStatus
+init_android_streams(PyThreadState *tstate)
+{
+    PyStatus status = _PyStatus_OK();
+    PyObject *_android_support = NULL;
+    PyObject *android_log_write = NULL;
+    PyObject *result = NULL;
+
+    _android_support = PyImport_ImportModule("_android_support");
+    if (_android_support == NULL)
+        goto error;
+
+    android_log_write = PyCFunction_New(&android_log_write_method, NULL);
+    if (android_log_write == NULL)
+        goto error;
+
+    // These log priorities match those used by Java's System.out and System.err.
+    result = PyObject_CallMethod(
+        _android_support, "init_streams", "Oii",
+        android_log_write, ANDROID_LOG_INFO, ANDROID_LOG_WARN);
+    if (result == NULL)
+        goto error;
+
+    goto done;
+
+error:
+    _PyErr_Print(tstate);
```

Comment:
It writes to stderr, which we just failed to initialize. It has also side effects which can be not desirable here. I suggest to remove this line.

Reply:
There will still be an stderr: the original one that writes to the C stdio. In fact, the [testbed app](https://github.com/python/cpython/pull/117878/files#diff-11524e84fc640f542ce2a0ba32bb833f8cc61bca9608b1fc279d16326a258965) redirects that to the system log as well, to allow us to see error messages from the early phases of the startup process. So this can still be useful.

What side effects could cause a problem?

Reply:
It sets `sys.last_exc` and calls `sys.excepthook()`. Unless the error is a `SystemExit`, in which case its behavior is more special. Is it okay here?

Reply:
I think so. This is a very early stage of initialization, so there can't be a custom excepthook. And the other side-effects are irrelevant because Python will exit immediately after.

I've tested this by adding deliberate errors to _android_support.py, and verifying that the traceback is printed to the C stderr, and that `Py_InitializeFromConfig` fails cleanly and returns the message "failed to initialize Android streams".